# Proyek Analisis Data: [Bike Sharing Dataset]
- **Nama:** [WAHYU DWI WICAKSONO]
- **Email:** [wahyudwiw20@gmail.com]
- **ID Dicoding:** [wahyu_dwi_wicaksono]

## Menentukan Pertanyaan Bisnis

## Pertanyaan Bisnis

1. Bagaimana pengaruh kategori suhu terhadap rata-rata jumlah penyewaan sepeda pada periode 2011–2012?

2. Pada jam berapa rata-rata jumlah penyewaan sepeda mencapai puncaknya dalam satu hari berdasarkan data tahun 2011–2012?
- Pertanyaan 1
- Pertanyaan 2

## Import Semua Packages/Library yang Digunakan

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Data Wrangling

### Gathering Data

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
df = pd.read_csv('/content/drive/MyDrive/Bike-sharing-dataset/day.csv')
df.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [4]:
df = pd.read_csv('/content/drive/MyDrive/Bike-sharing-dataset/hour.csv')
df.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


**Insight:**


*   Dataset terdiri dari data harian (day.csv) dan data per jam (hour.csv).
* Dataset memuat informasi jumlah penyewaan sepeda, cuaca, suhu, kelembapan, dan waktu.



### Assessing Data

In [7]:
day_df = pd.read_csv('/content/drive/MyDrive/Bike-sharing-dataset/day.csv')
hour_df = pd.read_csv('/content/drive/MyDrive/Bike-sharing-dataset/hour.csv')
day_df.info()
hour_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (tot

In [8]:
day_df.describe()

,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
mean,366.000000,2.496580,0.500684,6.519836,0.028728,2.997264,0.683995,1.395349,0.495385,0.474354,0.627894,0.190486,848.176471,3656.172367,4504.348837
std,211.165812,1.110807,0.500342,3.451913,0.167155,2.004787,0.465233,0.544894,0.183051,0.162961,0.142429,0.077498,686.622488,1560.256377,1937.211452
min,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000
25%,183.500000,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.337083,0.337842,0.520000,0.134950,315.500000,2497.000000,3152.000000
50%,366.000000,3.000000,1.000000,7.000000,0.000000,3.000000,1.000000,1.000000,0.498333,0.486733,0.626667,0.180975,713.000000,3662.000000,4548.000000
75%,548.500000,3.000000,1.000000,10.000000,0.000000,5.000000,1.000000,2.000000,0.655417,0.608602,0.730209,0.233214,1096.000000,4776.500000,5956.000000
max,731.000000,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000


In [9]:
hour_df.describe()

,instant,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,17379.0000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000
mean,8690.0000,2.501640,0.502561,6.537775,11.546752,0.028770,3.003683,0.682721,1.425283,0.496987,0.475775,0.627229,0.190098,35.676218,153.786869,189.463088
std,5017.0295,1.106918,0.500008,3.438776,6.914405,0.167165,2.005771,0.465431,0.639357,0.192556,0.171850,0.192930,0.122340,49.305030,151.357286,181.387599
min,1.0000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.020000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000
25%,4345.5000,2.000000,0.000000,4.000000,6.000000,0.000000,1.000000,0.000000,1.000000,0.340000,0.333300,0.480000,0.104500,4.000000,34.000000,40.000000
50%,8690.0000,3.000000,1.000000,7.000000,12.000000,0.000000,3.000000,1.000000,1.000000,0.500000,0.484800,0.630000,0.194000,17.000000,115.000000,142.000000
75%,13034.5000,3.000000,1.000000,10.000000,18.000000,0.000000,5.000000,1.000000,2.000000,0.660000,0.621200,0.780000,0.253700,48.000000,220.000000,281.000000
max,17379.0000,4.000000,1.000000,12.000000,23.000000,1.000000,6.000000,1.000000,4.000000,1.000000,1.000000,1.000000,0.850700,367.000000,886.000000,977.000000


In [10]:
day_df.isnull().sum()

,0
instant,0
dteday,0
season,0
yr,0
mnth,0
holiday,0
weekday,0
workingday,0
weathersit,0
temp,0


In [11]:
hour_df.isnull().sum()

,0
instant,0
dteday,0
season,0
yr,0
mnth,0
hr,0
holiday,0
weekday,0
workingday,0
weathersit,0


In [12]:
day_df.duplicated().sum()
hour_df.duplicated().sum()

np.int64(0)

**Insight:**
- Tidak ditemukan missing value yang signifikan pada dataset, sehingga data dapat langsung digunakan untuk proses analisis tanpa perlu penanganan khusus terhadap nilai kosong.
- Tidak terdapat data duplikat sehingga kualitas data cukup baik dan hasil analisis menjadi lebih akurat.
- Tipe data pada setiap kolom sudah sesuai, seperti kolom numerik untuk perhitungan dan kolom tanggal untuk analisis berbasis waktu.
- Dataset memiliki jumlah baris yang cukup banyak, sehingga representatif untuk menggambarkan pola penyewaan sepeda selama periode pengamatan.
- Secara keseluruhan, dataset berada dalam kondisi baik dan siap digunakan pada tahap pembersihan data serta eksplorasi lanjutan.

### Cleaning Data

In [13]:
day_df['dteday'] = pd.to_datetime(day_df['dteday'])
hour_df['dteday'] = pd.to_datetime(hour_df['dteday'])

In [14]:
hour_df['dteday'] = pd.to_datetime(hour_df['dteday'], format='%m/%d/%Y')

In [15]:
# kategori suhu
bins = [0, 0.25, 0.5, 0.75, 1]
labels = ['Dingin', 'Sedang', 'Hangat', 'Panas']
hour_df['temp_group'] = pd.cut(hour_df['temp'], bins=bins, labels=labels)

# kategori hari
hour_df['day_type'] = hour_df['workingday'].map({
    0: 'Akhir Pekan',
    1: 'Hari Kerja'
})

In [16]:
# mapping season
season_map = {
    1: 'Spring',
    2: 'Summer',
    3: 'Fall',
    4: 'Winter'
}
day_df['season'] = day_df['season'].map(season_map)
hour_df['season'] = hour_df['season'].map(season_map)

In [17]:
# mapping kondisi cuaca
weather_map = {
    1: 'Cerah',
    2: 'Berawan',
    3: 'Hujan Ringan',
    4: 'Hujan Lebat'
}
day_df['weathersit'] = day_df['weathersit'].map(weather_map)
hour_df['weathersit'] = hour_df['weathersit'].map(weather_map)

In [18]:
day_df.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,Spring,0,1,0,6,0,Berawan,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,Spring,0,1,0,0,0,Berawan,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,Spring,0,1,0,1,1,Cerah,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,Spring,0,1,0,2,1,Cerah,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,Spring,0,1,0,3,1,Cerah,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [19]:
hour_df.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt,temp_group,day_type
0,1,2011-01-01,Spring,0,1,0,0,6,0,Cerah,0.24,0.2879,0.81,0.0,3,13,16,Dingin,Akhir Pekan
1,2,2011-01-01,Spring,0,1,1,0,6,0,Cerah,0.22,0.2727,0.80,0.0,8,32,40,Dingin,Akhir Pekan
2,3,2011-01-01,Spring,0,1,2,0,6,0,Cerah,0.22,0.2727,0.80,0.0,5,27,32,Dingin,Akhir Pekan
3,4,2011-01-01,Spring,0,1,3,0,6,0,Cerah,0.24,0.2879,0.75,0.0,3,10,13,Dingin,Akhir Pekan
4,5,2011-01-01,Spring,0,1,4,0,6,0,Cerah,0.24,0.2879,0.75,0.0,0,1,1,Dingin,Akhir Pekan


**Insight:**
- Format kolom tanggal berhasil diubah ke tipe `datetime`, sehingga data dapat dianalisis berdasarkan hari, bulan, maupun tahun dengan lebih mudah.
- Perubahan tipe data tanggal juga mempermudah pembuatan filter rentang waktu pada dashboard Streamlit.
- Kolom kategori suhu (`temp_group`) ditambahkan agar nilai suhu numerik lebih mudah dipahami dalam bentuk kelompok seperti dingin, sedang, hangat, dan panas.
- Pengelompokan suhu membantu proses visualisasi sehingga hubungan antara suhu dan jumlah penyewaan dapat terlihat lebih jelas.
- Data yang telah dibersihkan dan ditambahkan fitur baru menjadi lebih siap untuk tahap Exploratory Data Analysis (EDA) dan visualisasi lanjutan.
- Proses cleaning ini meningkatkan kualitas dataset sehingga hasil analisis menjadi lebih terstruktur dan informatif.

## Exploratory Data Analysis (EDA)

### Explore ...

In [23]:
# Distribusi jumlah penyewaan
hour_df['cnt'].describe()

,cnt
count,17379.000000
mean,189.463088
std,181.387599
min,1.000000
25%,40.000000
50%,142.000000
75%,281.000000
max,977.000000


**Insight:**
- Rata-rata jumlah penyewaan menunjukkan permintaan yang cukup stabil.
- Terdapat beberapa waktu dengan jumlah penyewaan sangat tinggi dibanding rata-rata.

In [24]:
# Rata-rata penyewaan berdasarkan musim
hour_df.groupby(by="season")["cnt"].mean().reset_index()

,season,cnt
0,Fall,236.016237
1,Spring,111.114569
2,Summer,208.344069
3,Winter,198.868856


**Insight:**
- Terdapat perbedaan jumlah penyewaan pada setiap musim.
- Beberapa musim menunjukkan aktivitas penyewaan lebih tinggi dibanding musim lainnya.

In [25]:
# Rata-rata penyewaan berdasarkan hari kerja
hour_df.groupby(by="workingday")["cnt"].mean().reset_index()

,workingday,cnt
0,0,181.405332
1,1,193.207754


**Insight:**
- Suhu memiliki hubungan positif terhadap jumlah penyewaan.
- Kelembapan dan kecepatan angin cenderung memiliki hubungan negatif terhadap penyewaan.
- Faktor lingkungan berpengaruh terhadap keputusan pengguna dalam menyewa sepeda

## Visualization & Explanatory Analysis

### Pertanyaan 1: ### Bagaimana pengaruh kategori suhu terhadap rata-rata jumlah penyewaan sepeda pada periode 2011–2012?

In [26]:
# Membuat salinan data
temp_df = day_df.copy()

# Membuat kategori suhu
bins = [0, 0.4, 0.7, 1]
labels = ['Sangat Dingin', 'Nyaman', 'Panas']

temp_df['temp_category'] = pd.cut(
    temp_df['temp'],
    bins=bins,
    labels=labels,
    include_lowest=True
)

# Melihat jumlah data tiap kategori
temp_df['temp_category'].value_counts()

,count
temp_category,
Nyaman,341
Sangat Dingin,261
Panas,129


In [27]:
# Rata-rata penyewaan tiap kategori suhu
temp_analysis = temp_df.groupby(
    by="temp_category",
    observed=True
)["cnt"].mean().reset_index()

temp_analysis

,temp_category,cnt
0,Sangat Dingin,2965.517241
1,Nyaman,5243.598240
2,Panas,5663.658915


**Insight:**
- Kategori suhu **Nyaman** memiliki jumlah data terbanyak sebanyak 341 observasi, menunjukkan sebagian besar hari berada pada kondisi suhu sedang dan nyaman.
- Kategori suhu **Sangat Dingin** memiliki 261 observasi, sedangkan kategori **Panas** memiliki jumlah observasi paling sedikit yaitu 129.
- Rata-rata jumlah penyewaan sepeda tertinggi berada pada kategori **Panas** dengan nilai sekitar 5663,66 penyewaan.
- Kategori **Nyaman** berada di urutan kedua dengan rata-rata sekitar 5243,60 penyewaan.
- Kategori **Sangat Dingin** memiliki rata-rata penyewaan terendah yaitu sekitar 2965,52.
- Hasil ini menunjukkan bahwa semakin hangat suhu udara, kecenderungan masyarakat untuk menyewa sepeda semakin meningkat.
- Temuan ini dapat menjadi dasar dalam pengambilan keputusan operasional dan strategi promosi layanan penyewaan sepeda.

### Pertanyaan 2: ### Pada jam berapa rata-rata jumlah penyewaan sepeda mencapai puncaknya dalam satu hari berdasarkan data tahun 2011–2012?

In [28]:
# Rata-rata penyewaan per jam
hour_analysis = hour_df.groupby(by="hr")["cnt"].mean().reset_index()

# Urutkan dari tertinggi
hour_analysis.sort_values(by="cnt", ascending=False).head(10)

,hr,cnt
17,17,461.452055
18,18,425.510989
8,8,359.011004
16,16,311.983562
19,19,311.523352
13,13,253.661180
12,12,253.315934
15,15,251.233196
14,14,240.949246
20,20,226.030220


In [29]:
# Membuat kategori waktu
def get_time_category(hour):
    if 6 <= hour <= 10:
        return 'Morning Peak'
    elif 11 <= hour <= 15:
        return 'Mid-day'
    elif 16 <= hour <= 20:
        return 'Evening Peak'
    else:
        return 'Off-Peak/Night'

hour_df['time_category'] = hour_df['hr'].apply(get_time_category)

# Rata-rata penyewaan berdasarkan kategori waktu
time_analysis = hour_df.groupby(
    by="time_category"
)["cnt"].mean().reset_index()

time_analysis

,time_category,cnt
0,Evening Peak,347.343304
1,Mid-day,241.475563
2,Morning Peak,208.092210
3,Off-Peak/Night,60.604799


**Insight:**
- Jam dengan rata-rata penyewaan tertinggi terjadi pada **pukul 17.00** dengan nilai sekitar **461,45**, diikuti **pukul 18.00** sebesar **425,51**.
- Peningkatan penyewaan juga terlihat pada **pukul 08.00** dengan rata-rata sekitar **359,01**, yang menunjukkan aktivitas pagi hari cukup tinggi.
- Kategori waktu **Evening Peak** memiliki rata-rata penyewaan tertinggi yaitu sekitar **347,34**, sehingga sore hari merupakan waktu paling sibuk.
- Kategori **Mid-day** berada di urutan kedua dengan rata-rata sekitar **241,48**, diikuti **Morning Peak** sebesar **208,09**.
- Kategori **Off-Peak/Night** memiliki rata-rata penyewaan paling rendah yaitu sekitar **60,60**, menunjukkan aktivitas malam hari relatif sedikit.
- Pola ini menunjukkan bahwa sepeda banyak digunakan sebagai sarana mobilitas pada jam berangkat dan pulang kerja atau sekolah.
- Informasi ini dapat dimanfaatkan untuk pengaturan jumlah sepeda dan strategi operasional pada jam-jam sibuk.

## Analisis Lanjutan (Opsional)

In [32]:
import os
from google.colab import files

os.makedirs("dashboard", exist_ok=True)

hour_df.to_csv("dashboard/main_data.csv", index=False)

files.download("dashboard/main_data.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Conclusion

Conclusion Pertanyaan 1

Kategori suhu memiliki pengaruh terhadap jumlah penyewaan sepeda. Berdasarkan hasil analisis, kategori suhu Panas memiliki rata-rata penyewaan tertinggi, diikuti kategori Nyaman, sedangkan kategori Sangat Dingin memiliki jumlah penyewaan paling rendah. Hal ini menunjukkan bahwa kondisi cuaca yang lebih hangat meningkatkan minat masyarakat dalam menggunakan layanan penyewaan sepeda.

Conclusion Pertanyaan 2

Rata-rata jumlah penyewaan sepeda mencapai puncaknya pada pukul 17.00, diikuti pukul 18.00 dan 08.00. Jika dikelompokkan berdasarkan kategori waktu, periode Evening Peak menjadi waktu tersibuk. Temuan ini menunjukkan bahwa sepeda banyak digunakan untuk mobilitas saat jam pulang dan berangkat kerja atau sekolah.